### PR TITLE
kubectl: Update printers to use an external object instead of internal

### PR DESCRIPTION
**Pull request by seans3**
* Updates `get` to fall back to printing external objects instead of internal objects (server-side printing now default, so this happens when server-side printing fails).
* Copies internal printers (printers.go) into `kubectl` and changes all object printing functions to take external objects instead of internal objects.
* Copies the human readable printer into `kubectl`.

Helps fix:
https://github.com/kubernetes/kubectl/issues/80

```release-note
NONE
```

/kind cleanup
/area kubectl
/sig cli
/assign

### CODE REVIEW

File: pkg/kubectl/cmd/get/humanreadable.go [hunk 1]
```diff
@@ -0,0 +1,874 @@
...
+	w, found := output.(*tabwriter.Writer)
+	if !found && !h.skipTabWriter {
+		w = utilprinters.GetNewTabWriter(output)
+		output = w
+		defer w.Flush()
+	}
+
+	// display tables following the rules of options
+	if table, ok := obj.(*metav1beta1.Table); ok {
+		if err := DecorateTable(table, h.options); err != nil {
```
Comment (liggitt) [inline pkg/kubectl/cmd/get/humanreadable.go]: is this replicating table-printing code? that part seems like it should move to cli-runtime. I expected the only duplication in this PR to be things we could completely delete after 1.16, once we have support for server-side printing of watch events

Comment (liggitt) [reply pkg/kubectl/cmd/get/humanreadable.go]: moving the existing DecorateTable and PrintTable to cli-runtime/pkg/printers/table.go would make sense to me

cc @smarterclayton for plans for collapsing onto server-side printing

File: pkg/kubectl/cmd/get/humanreadable.go [hunk 2]
```diff
@@ -0,0 +1,874 @@
...
+
+	table.ColumnDefinitions = columns
+	table.Rows = rows
+	return nil
+}
+
+// PrintTable returns a table for the provided object, using the printer registered for that type. It returns
+// a table that includes all of the information requested by options, but will not remove rows or columns. The
+// caller is responsible for applying rules related to filtering rows or columns.
+func (h *HumanReadablePrinter) PrintTable(obj runtime.Object, options PrintOptions) (*metav1beta1.Table, error) {
```
Comment (liggitt) [inline pkg/kubectl/cmd/get/humanreadable.go]: we shouldn't copy any functions that are not used... I think this is only used server-side

**Comment by liggitt** [comment]
the lifetime of these copied typed handlers is just for 1.15, right? as of 1.16, we can rely on server-side printing in kubectl and delete these? this seems like a lot of code motion just to have for a single release... is it actually unblocking a lot of work sig-cli plans to do in 1.15?

if it's considered worthwhile, let's structure this so that the code intended to be deleted in 1.16 is very clearly marked, and the code intended to survive is well-factored (e.g. https://github.com/kubernetes/kubernetes/pull/75995#discussion_r271119146)

**Comment by liggitt** [comment]
/hold

to make sure we avoid duplicating the parts of humanreadableprinter that will survive past 1.15

File: pkg/kubectl/cmd/get/humanreadable.go (hunk 1 above)
Comment (seans3) [reply pkg/kubectl/cmd/get/humanreadable.go]: I believe a significant portions of `humanreadable.go` will be used by both `kubectl` and `registry` after 1.15. In that case, the best way forward may be to just move all of `humanreadable.go` (and `humanreadable_test.go` as well as `PrintOptions`) into a common location out of core: `cli-runtime`. The copied external type handlers (`printer.go`) will be clearly marked as removable after 1.15 and can be moved to a subdirectory of `get` called `legacy` or renamed `legacy_printers.go`. Let me know what you think.

Comment (liggitt) [reply pkg/kubectl/cmd/get/humanreadable.go]: moving the bits needed to print tables to cli-runtime makes sense to me

xref https://github.com/kubernetes/kubernetes/issues/74179

https://github.com/kubernetes/kubernetes/pull/76161 allows kubectl to use server-side printing for all get/watch requests

https://github.com/kubernetes/kubernetes/pull/76272 removes unused methods from humanreadable printer and removes internal decoding from it (if decoding to internal is required, it should be done before passing to PrintObj, like kubectl get currently does)

**Comment by liggitt** [comment]
closing in favor of https://github.com/kubernetes/kubernetes/pull/79530

/close

**Pull request closed without merging**